Object-in-fluid implementation to python

icimrak opened the pull request:
Fixes #

Description of changes:
 - 


PR Checklist
------------
 - [ ] Tests?
   - [ ] Interface
   - [ ] Core 
 - [ ] Docs?

icimrak committed:
Added calculation of ks_pressure to the oif_object_analyze function of the scripts/object_in_fluid.tcl

icimrak committed:
minor bug corrected

icimrak committed:
oif: When mirroring in creating the template, it is possible to output corrected mesh and incidence files Nodes.dat and Triangles dat

icimrak committed:
Bug in template mirroring has been corrected. oif_create_template mirror

icimrak committed:
Some checks added. Program will exit if not all mandatory arguments are given for oif_.... commands

icimrak committed:
added edge-statistics to object_in_fluid.tcl

icimrak committed:
Minor bug in oif_object_analyze solved

icimrak committed:
Implementatio of proportional local area forces into oif_local_forces, according to artcile:
Jancigova, Cimrak,Non-uniform force allocation for area preservation in spring network models,
		International Journal for Numerical Methods in Biomedical Engineering, DOI: 10.1002/cnm.2757

icimrak committed:
Global area was updated so that forces are ditributed proportionally, according to the article
		I.Jancigova, I.Cimrak,
		Non-uniform force allocation for area preservation in spring network models,
		International Journal for Numerical Methods in Biomedical Engineering, DOI: 10.1002/cnm.2757

Global volume forces remain to be adapted to fulfill force-free condition. To be done in later commits.

icimrak committed:
OIF transition to python

RudolfWeeber committed:
Fix for oi bonded interactions

icimrak committed:
somethond added to interactions for oif python transition

RudolfWeeber committed:
Merge branch 'python' of git://github.com/icimrak/espresso into object_in_fluid

icimrak committed:
some advancement in python

RudolfWeeber committed:
Fix oif related interactions and particle properties, testcase for particle properties

icimrak committed:
in object_in_fluid, template for von-misses stress was created to oif_object_analyze

icimrak committed:
Implemented computations of pricipal directions of rotated and deformed cells. implemented in oif_object_analyze. It is though slow....

icimrak committed:
Added solid feature for OIF template....

icimrak committed:
minor additions to checkpointing in oif python implementation

icimrak committed:
Merge remote-tracking branch 'espressomd_master/python' into python

icimrak committed:
Modified global area force by adding the weghting S_triangle/S_wholecell. To avoid merging of mesh points....

Commit:
After the code did not compile I changed the path to reaction.pyx and then it worked out....

Commit:
feature mirror removed from oif_create_template. Instead, swith mirror was added to oif_mesh_analyze, that takes the mesh files and mirrored them.

Commit:
Merge branch 'tmp_newglobal' into oif_periodic

Commit:
Viscosity implemented according to M.Busik sources

Commit:
Added particle feature mol to python interface.

Commit:
Merge branch 'python' into espmd_python

Commit:
Added python interface for out_direction bonded interaction

Commit:
Viscosity corrected by rescaling, now the viscosity coefficient is independet of the time step.

Commit:
Merge branch 'espmd_python' into python

Commit:
Merge branch 'viscosity' into python
'viscosity' was a short branch from 'oif_periodic'. This means, that all work done on oif_periodic is now merged into current 'python' development branch, including new implementation for global area forces, viscosity....

Commit:
fixed oif local area calculation according to the paper DOI: 10.1002/cnm.2757

Commit:
added soft sphere and fixed membrane collision

Commit:
calculation of global area force edited for better understanding

Commit:
Variant B of viscosity impleentation was changed to correct projection onto the line btw mesh points

Commit:
Merge branch 'viscosity' into python

Commit:
Merge remote-tracking branch 'remotes/ijancigova/python' into python
There is soft-sphere and membrane-collision included

icimrak committed:
Merge remote-tracking branch 'espressomd_master/python' into python
Conflicts resolved, however there are different results in simulations compared to previous commit.

Problems resolved...

icimrak committed:
Merge remote-tracking branch 'espressomd_master/python' into python

icimrak committed:
Two source files added for python implementation of object-in-fluid module

RudolfWeeber commented:
Thanks alot. It is gret to have a Python implementation for the method.
I had a 1st look and found the following issues.

* We are using pep8 naming convention in Python. Classes are CamelCase with the 1st letter a capital. Method and variable names are lower_case_underscored.
* Please put docstrings on the methods in your classes. We use numpydoc conventions
* Please perform type checking on the arguments passed to the constructors of your classes.
  You can use isinstance(), e.g., on arguments of type ParticleHandle and
  and check_type_or_throw_except() from espressomd.utils. You will need to replace 
  cdef check_type_type_or_throw_except()
  by cpdef 
  in utils.pyx to make it accessibel from pure .py files.
* Please use numpy arrays rather than lists or even individual x, y, and z-variables. This makes all the math much more straight forward and removes code repetition for different coordinates.
* The 
  from oif_utils import * 
  leads to the siplling of methods with rather general names into the oif_classes namespace.
* Please 
  raise Exception("Message")
  rather than printing somethign and then doing quit()
* Edge class: Your distance calculation does not respect periodic boundary conditions  (minimum image convention).
  For two particles, you can use the distance() and distance_vec() methods from the espressomd.system.System class.
* There seems to be a lot of io code. Is it possible to use the pickle-Python module and implementing __get_state__ and __set_state__
* Rather than using functions that print() something, please return the state ina data structure
* If you need string representations for objects so you can do
  print(cell)
  please implement the __repr__ or __str__ method on the class
* Could you please provide a sample sample script, which illustrates, how to use oif?

RudolfWeeber commented:
Regarding docs.
We are switching to Sphinx-based docs.
The idea, is to put  high-level description (and sometimes an example) in
src/sphinx/oif.rst
and docstrings describing the purpose methods and attributes of classes in the in-code doc via docstrings. The .tex-Doc is no longer used.
There are tools to automtically convert from .tex to .rst, e.g. pandoc.
The current state of the sphinx-based doc is always available at
espressomd.org/html/doc

To build it for yourself, use
make sphinx
from the top build dir. The doc will be in doc/sphinx/html.

icimrak committed:
OIF documentation updated

icimrak committed:
DOCSTRINGS for object-in-fluid added

Commit:
applied pep8 convention to oif_classes.py and oif_utils.py

icimrak committed:
som minor things in oif-python files

icimrak committed:
Merge remote-tracking branch 'ijancigova/python_current' into python

pep8 conversion done

icimrak committed:
Type checking added to python implementation of oif classes

Commit:
oif: changed calculations with coordinates to numpy vector calculations where applicable

icimrak committed:
python 2 version of print comand changed to python 2 and 3 version

icimrak committed:
Merge remote-tracking branch 'espressomd_master/python' into python

Commit:
some printouts changed to compatible with python 3
som "is" changed to "==" in if statements

KaiSzuttor commented:
Please use an indentation level of 4 spaces.

RudolfWeeber commented:
> Please use an indentation level of 4 spaces.
autopep8 -i file.py

Commit:
Minor fixes in object-in-fluid: replacing pointwise operations using numpy, removing unnecessary linebreaks, corrected retyping of data input from files

Commit:
Renormalization by phi0 removed from the computations of bending forces, to be consistent with Fedosov and Krueger

Commit:
global area force changed to conform the "energy" approach, division by total surface added....

KaiSzuttor commented on src/python/espressomd/scripts/oif_classes.py:
```diff
@@ -942,14 +942,14 @@ def append_point_data_to_vtk(self, file_name=None, data_name=None, data=None, fi
     def output_raw_data(self, file_name=None, data=None):
         if file_name is None:
-            print "OifCell: output_raw_data: No file_name provided."
+            print ("OifCell: output_raw_data: No file_name provided.")
```
Please don't use prints but raise Exceptions

KaiSzuttor commented on src/core/object-in-fluid/oif_global_forces.cpp:
```diff
@@ -340,64 +343,59 @@ void add_oif_global_forces(double *area_volume,
+        VOL_vv=(VOL_volume - iaparams->p.oif_global_forces.V0)/iaparams->p.oif_global_forces.V0;          
+        for(k=0;k<3;k++) {
+          VOL_force[k]=iaparams->p.oif_global_forces.kv * VOL_vv * VOL_A * VOL_norm[k]/VOL_dn * 1.0 / 3.0;
+          //printf("%e ",force[k]);
```
please remove commented code

KaiSzuttor replied:
pls remove

KaiSzuttor commented on src/core/object-in-fluid/oif_local_forces.hpp:
```diff
@@ -116,7 +120,7 @@ inline int calc_oif_local(Particle *p2, Particle *p1, Particle *p3, Particle *p4
                     }
 				} else {
-					printf("Something wrong in oif_local_forces.hpp: All particles in a bond are ghost particles, impossible to unfold the positions...");
+					printf("Something wrong in oif_local_forces.hpp: All particles in a bond are ghost particles, impossible to unfold the positions...\n");
```
please raise Exception here

KaiSzuttor commented on src/core/object-in-fluid/oif_local_forces.hpp:
```diff
@@ -157,6 +161,78 @@ inline int calc_oif_local(Particle *p2, Particle *p1, Particle *p3, Particle *p4
+        len2 = sqrlen(dx);
+        len = sqrt(len2);
+          
+		//len_new = sqrt(pow((fp2[0] +  p2->m.v[0]) - (fp3[0] +  p3->m.v[0]),2) + pow((fp2[1] +  p2->m.v[1]) - (fp3[1] +  p3->m.v[1]),2) + pow((fp2[2] +  p2->m.v[2]) - (fp3[2] +  p3->m.v[2]),2));
```
please remove commented out code

KaiSzuttor commented on src/core/object-in-fluid/oif_local_forces.hpp:
```diff
@@ -169,7 +245,8 @@ inline int calc_oif_local(Particle *p2, Particle *p1, Particle *p3, Particle *p4
         dn2=normr(n2);
         phi = angle_btw_triangles(fp1,fp2,fp3,fp4);
         if (phi < 0.001 || phi > 2*M_PI - 0.001) printf("oif_local_forces.hpp, calc_oif_local: Angle approaches 0 or 2*Pi\n");
```
remote debug printf

KaiSzuttor commented on src/python/espressomd/interactions.pxd:
```diff
@@ -245,6 +281,7 @@ cdef extern from "interaction_data.hpp":
         Fene_bond_parameters fene
         Oif_global_forces_bond_parameters oif_global_forces
         Oif_local_forces_bond_parameters oif_local_forces
+        #Oif_out_direction_bond_parameters oif_out_direction
```
remove commented out code

KaiSzuttor commented on src/python/espressomd/scripts/oif_classes.py:
```diff
@@ -0,0 +1,1294 @@
+        self.ids_extremal_points = [0, 0, 0, 0, 0, 0, 0]
+        
+        if (nodes_file is None) or (triangles_file is None):
+            print("OifMesh warning: one of nodes_file or triangles_file was not given. "
```
raise Exception

KaiSzuttor commented on src/python/espressomd/scripts/oif_classes.py:
```diff
@@ -0,0 +1,1294 @@
+
+            if check_orientation is True:
+                # check whether all triangles in file had the same orientation; if not, correct the orientation
+                print("OifMesh: Checking orientation of triangles (and repairing if needed). "
```
raise Exception

KaiSzuttor commented on src/python/espressomd/scripts/oif_classes.py:
```diff
@@ -0,0 +1,1294 @@
+                            tmp_triangle_list_ok.append(tmp_triangle)
+                        else:
+                            tmp_triangle_list_ok.append(corrected_triangle)
+                            print("OifMesh: Correcting orientation of triangle.")
```
remove print

KaiSzuttor commented on src/python/espressomd/scripts/oif_classes.py:
```diff
@@ -0,0 +1,1294 @@
+            for tmp_triangle in self.triangles:
+                    self.triangles[i] = Triangle(tmp_triangle.A, tmp_triangle.C, tmp_triangle.B)
+                    i += 1
+        print("OifMesh: Triangulation correct.")
```
remove print

KaiSzuttor commented on src/python/espressomd/scripts/oif_classes.py:
```diff
@@ -0,0 +1,1294 @@
+        if not isinstance(check_orientation,bool):
+            raise TypeError("check_orientation must be bool.")     
+        if normal is False:
+            print("OifCellType warning: Option normal is not used => membrane collision will not work.")
```
raise Exception

KaiSzuttor commented on src/python/espressomd/scripts/oif_classes.py:
```diff
@@ -0,0 +1,1294 @@
+        if normal is False:
+            print("OifCellType warning: Option normal is not used => membrane collision will not work.")
+        if check_orientation is False:
+            print("OifCellType warning: Check of orientation of triangles is switched off.")
```
remove print

KaiSzuttor commented on src/python/espressomd/scripts/oif_classes.py:
```diff
@@ -0,0 +1,1294 @@
+                p.part.add_bond((tmp_out_direction_interaction, self.mesh.neighbors[p.id].A.part_id,
+                                 self.mesh.neighbors[p.id].B.part_id, self.mesh.neighbors[p.id].C.part_id))
+
+        self.print_info()
```
I think it's not a good idea to default print all info

KaiSzuttor commented on src/python/espressomd/scripts/oif_classes.py:
```diff
@@ -0,0 +1,1294 @@
+
+    def output_vtk_pos(self, file_name=None):
+        if file_name is None:
+            print("OifCell: No file_name provided for vtk output.")
```
raise Exception

KaiSzuttor commented on src/python/espressomd/scripts/oif_classes.py:
```diff
@@ -0,0 +1,1294 @@
+            print("OifCell: No file_name provided for vtk output.")
+            return
+        if ".vtk" not in file_name:
+            print("OifCell warning: A file with vtk format will be written without .vtk extension.")
```
remove print

KaiSzuttor commented on src/python/espressomd/scripts/oif_classes.py:
```diff
@@ -0,0 +1,1294 @@
+
+    def output_vtk_pos_folded(self, file_name=None):
+        if file_name is None:
+            print("OifCell: No file_name provided for vtk output.")
```
raise Exception

KaiSzuttor commented on src/python/espressomd/scripts/oif_classes.py:
```diff
@@ -0,0 +1,1294 @@
+            print("OifCell: No file_name provided for vtk output.")
+            return
+        if ".vtk" not in file_name:
+            print("OifCell warning: A file with vtk format will be written without .vtk extension.")
```
remove print

KaiSzuttor commented on src/python/espressomd/scripts/oif_classes.py:
```diff
@@ -0,0 +1,1294 @@
+
+    def append_point_data_to_vtk(self, file_name=None, data_name=None, data=None, first_append=None):
+        if file_name is None:
+            print("OifCell: append_point_data_to_vtk: No file_name provided.")
```
raise Exception

KaiSzuttor commented on src/python/espressomd/scripts/oif_classes.py:
```diff
@@ -0,0 +1,1294 @@
+            print("OifCell: append_point_data_to_vtk: No file_name provided.")
+            return
+        if data is None:
+            print("OifCell: append_point_data_to_vtk: No data provided.")
```
raise Exception

KaiSzuttor commented on src/python/espressomd/scripts/oif_classes.py:
```diff
@@ -0,0 +1,1294 @@
+            print("OifCell: append_point_data_to_vtk: No data provided.")
+            return
+        if data_name is None:
+            print("OifCell: append_point_data_to_vtk: No data_name provided.")
```
raise

KaiSzuttor commented on src/python/espressomd/scripts/oif_classes.py:
```diff
@@ -0,0 +1,1294 @@
+            print("OifCell: append_point_data_to_vtk: No data_name provided.")
+            return
+        if first_append is None:
+            print("OifCell: append_point_data_to_vtk: Need to know whether this is the first data list to be "
```
raise

KaiSzuttor commented on src/python/espressomd/scripts/oif_classes.py:
```diff
@@ -0,0 +1,1294 @@
+            return
+        n_points = self.get_n_nodes()
+        if (len(data) != n_points):
+            print ("OifCell: append_point_data_to_vtk: Number of data points does not match number of mesh points.")
```
raise

KaiSzuttor commented on src/python/espressomd/scripts/oif_classes.py:
```diff
@@ -0,0 +1,1294 @@
+
+    def output_raw_data(self, file_name=None, data=None):
+        if file_name is None:
+            print ("OifCell: output_raw_data: No file_name provided.")
```
raise

KaiSzuttor commented on src/python/espressomd/scripts/oif_classes.py:
```diff
@@ -0,0 +1,1294 @@
+            print ("OifCell: output_raw_data: No file_name provided.")
+            return
+        if data is None:
+            print ("OifCell: output_raw_data: No data provided.")
```
raise

KaiSzuttor commented on src/python/espressomd/scripts/oif_classes.py:
```diff
@@ -0,0 +1,1294 @@
+            return
+        n_points = self.get_n_nodes()
+        if (len(data) != n_points):
+            print ("OifCell: output_raw_data: Number of data points does not match number of mesh points.")
```
raise

KaiSzuttor commented on src/python/espressomd/scripts/oif_classes.py:
```diff
@@ -0,0 +1,1294 @@
+
+    def output_mesh_points(self, file_name=None):
+        if file_name is None:
+            print ("OifCell: No file_name provided for mesh nodes output.")
```
raise

KaiSzuttor commented on src/python/espressomd/scripts/oif_classes.py:
```diff
@@ -0,0 +1,1294 @@
+
+    def set_mesh_points(self, file_name=None):
+        if file_name is None:
+            print ("OifCell: No file_name provided for set_mesh_points. ")
```
raise

KaiSzuttor commented on src/python/espressomd/scripts/oif_classes.py:
```diff
@@ -0,0 +1,1294 @@
+
+        for i in range(0,6):
+            if (el_forces[i] != 0) and (el_forces[i] != 1):
+                print("OifCell: elastic_forces: Incorrect argument. el_forces has to be a sixtuple of 0s and 1s, "
```
raise

KaiSzuttor commented on src/python/espressomd/scripts/oif_classes.py:
```diff
@@ -0,0 +1,1294 @@
+        for i in range(0,6):
+            if (f_metric[i] != 0) and (f_metric[i] != 1):
+                print("OifCell: elastic_forces: Incorrect argument. f_metric has to be a sixtuple of 0s and 1s, "
+                      "specifying which f_metric will be calculated. The order in the sixtuple is (ks, kb, kal, "
```
raise

KaiSzuttor commented on src/python/espressomd/scripts/oif_classes.py:
```diff
@@ -0,0 +1,1294 @@
+        # output vtk (folded)
+        if vtk_file is not None:
+            if el_forces == (0, 0, 0, 0, 0, 0):
+                print("OifCell: elastic_forces: The option elastic_forces was not used. "
```
remove print

KaiSzuttor commented on src/python/espressomd/scripts/oif_classes.py:
```diff
@@ -0,0 +1,1294 @@
+        # output raw data
+        if raw_data_file is not None:
+            if (el_forces[0] + el_forces[1] + el_forces[2] + el_forces[3] + el_forces[4] + el_forces[5]) != 1:
+                print("OifCell: elastic_forces: Only one type of elastic forces can be written into one "
```
raise

KaiSzuttor commented on src/python/espressomd/scripts/oif_utils.py:
```diff
@@ -0,0 +1,406 @@
+
+
+def get_lb_interpolated_velocity(position, lbf, system, fluid_agrid):
+    # position is a vector [x,y,z], in which we want the fluid velocity
```
please add numpy-style documentation

KaiSzuttor commented on src/python/espressomd/scripts/oif_utils.py:
```diff
@@ -0,0 +1,406 @@
+
+def output_vtk_lines(lines, out_file):
+    """
+    Outputs the VTK files for visualisation of lines in e.g. Paraview.
```
please add a parameters section in numpy-doc format

KaiSzuttor commented on src/python/espressomd/scripts/oif_utils.py:
```diff
@@ -0,0 +1,406 @@
+
+def output_vtk_cylinder(center, axis, length, radius, n, out_file):
+    """
+    Outputs the VTK files for visualisation of a cylinder in e.g. Paraview.
```
please add parameters section in numpydoc format

KaiSzuttor commented on src/python/espressomd/scripts/oif_utils.py:
```diff
@@ -0,0 +1,406 @@
+
+def output_vtk_rhomboid(corner, a, b, c, out_file):
+    """
+    Outputs the VTK files for visualisation of a rhomboid in e.g. Paraview.
```
please add parameters section in numpydoc format

KaiSzuttor commented on src/python/espressomd/scripts/oif_utils.py:
```diff
@@ -0,0 +1,406 @@
+
+def oif_calc_volume_force(kv, pA, pB, pC, V0, V):
+    """
+    Calculates volume forces for three points in a triangle.
```
please add parameters section in numpydoc format

KaiSzuttor commented on src/python/espressomd/scripts/oif_utils.py:
```diff
@@ -0,0 +1,406 @@
+
+def oif_calc_global_area_force(kag, pA, pB, pC, Ag0, Ag):
+    """
+    Calculates global area forces between three points in a triangle.
```
please add parameters section in numpydoc format

KaiSzuttor commented on src/python/espressomd/scripts/oif_utils.py:
```diff
@@ -0,0 +1,406 @@
+
+def oif_calc_local_area_force(kal, pA, pB, pC, A0, A):
+    """
+    Calculates local area forces between three points in one triangle.
```
please add parameters section in numpydoc format

KaiSzuttor commented on src/python/espressomd/scripts/oif_utils.py:
```diff
@@ -0,0 +1,406 @@
+
+def oif_calc_bending_force(kb, pA, pB, pC, pD, phi0, phi):
+    """
+    Calculates bending forces for four points on two adjacent triangles.
```
please add parameters section in numpydoc format

KaiSzuttor commented on src/python/espressomd/scripts/oif_utils.py:
```diff
@@ -0,0 +1,406 @@
+
+def oif_calc_linear_stretching_force(ks, pA, pB, dist0, dist):
+    """
+    Calculates linear stretching forces between two points on an edge.
```
please add parameters section in numpydoc format

KaiSzuttor commented on src/python/espressomd/scripts/oif_utils.py:
```diff
@@ -0,0 +1,406 @@
+
+def oif_calc_stretching_force(ks, pA, pB, dist0, dist):
+    """
+    Calculates nonlinear stretching forces between two points on an edge.
```
please add parameters section in numpydoc format

KaiSzuttor commented on src/python/espressomd/scripts/oif_utils.py:
```diff
@@ -0,0 +1,406 @@
+
+def oif_neo_hookean_nonlin(lambd):
+    """
+    Defines NeoHookean nonlinearity.
```
please add parameters section in numpydoc format

KaiSzuttor commented on src/python/espressomd/scripts/oif_utils.py:
```diff
@@ -0,0 +1,406 @@
+
+def discard_epsilon(x):
+    """
+    Returns zero if the argument is too small.
```
please add parameters section in numpydoc format

KaiSzuttor commented on src/python/espressomd/scripts/oif_utils.py:
```diff
@@ -0,0 +1,406 @@
+
+def angle_btw_triangles(P1, P2, P3, P4):
+    """
+    Returns the size of an angle between triangles given by points P2, P1, P3 and P2, P3, P4.
```
please add parameters section in numpydoc format

KaiSzuttor commented on src/python/espressomd/scripts/oif_utils.py:
```diff
@@ -0,0 +1,406 @@
+
+def vec_distance(a, b):
+    """
+    Returns the length of vector between points a and b.
```
please add parameters section in numpydoc format

KaiSzuttor commented on src/python/espressomd/scripts/oif_utils.py:
```diff
@@ -0,0 +1,406 @@
+
+def norm(vect):
+    """
+    Returns the norm of a vector.
```
please add parameters section in numpydoc format

KaiSzuttor commented on src/python/espressomd/scripts/oif_utils.py:
```diff
@@ -0,0 +1,406 @@
+
+small_epsilon = 0.000000001
+large_number = 10000000.0
+pi = 3.1415926535897931
```
please use np.pi

RudolfWeeber commented:
Once Kai's comments are adressed, I suggest, we integrate this such that pypresso provides a separate package "object_in_fluid". I.e., we move the "scripts" folder to src/python/object_in_fluid and provide an __init__.py to load the relevant stuff into the top module name space.

Commit:
Clean up of the code:
- commented out code removed
- Exceptions raised instead of printouts
- debug printouts removed
- np.pi used
- parameters section in numpydoc format added to utils.py

icimrak commented:
All Kai's comments resolved...

icimrak commented:
I agree with Rudolf's suggestion....

RudolfWeeber committed:
Merge branch 'python' of git://github.com/espressomd/espresso into oif_merge

RudolfWeeber committed:
oif: fixes and sample script

RudolfWeeber commented:
So, I merged the current development branch and sorted out the regressions.
Based on the previous discussion, oif is now in a separate Python module

./pypresso
>>> import object_in_fluid
The sources are at src/python/object_in_fluid


I also included thee motivation.py from the website in samples/object_in_fluid and got it running.

Can you please check, if it does what it is supposed to do?
The best solution would be a script that loads an object (sphere...), deforms it and checks if the original shape is restored. This can then be run automatically.

The current documentation (doc/sphinx/oif.rst) does not reflect the pep8-naming now applied in the oif module. Also, some name changes (stretching -> resize) are not reflected. Can you please adapt the docs
Also, the docs in doc/sphinx/inter.rst need updating (membrane_collision, ...). Currently, the TCL stat is described.

RudolfWeeber committed:
oif: missing file

RudolfWeeber committed:
oif: new-style classes

RudolfWeeber committed:
oif: Remove GHOST_FLAG dependency.

As this is no longer a switchable feature. Kept the GHOST_FLAGS enabled code.

RudolfWeeber committed:
oif: rm sample output

icimrak commented:
Ok, thanks. I will do all the suggestions, however I will not have time this week. Beginning next week the work will be done...

Ivan

Commit:
documentation of OIF adapted to python interface instead of TCL

Commit:
Added unit test into samples/object_in_fluid/unit_test.py

KaiSzuttor commented on samples/object_in_fluid/unit-test.py:
```diff
@@ -0,0 +1,70 @@
+import numpy as np
+import os, sys
+
+from object_in_fluid import *
```
a wildcard import is not a good choice due to namespace pollution and readability, maybe sth. like the following would be appropriate:
```python
import object_in_fluid as oif
```

Commit:
unit test for oif in samples/ was slightly changed

Commit:
sample script motivation.py in samples has been slightly changed

icimrak commented:
The changes requested by Rudolf and Kai has been implemented:

Rudolf: I also included the motivation.py from the website in samples/object_in_fluid and got it running. Can you please check, if it does what it is supposed to do?
Response IC: I slightly changed this script. It is aimed as a starting point for a simulation of cell flow.

Rudolf: The best solution would be a script that loads an object (sphere...), deforms it and checks if the original shape is restored. This can then be run automatically.
Response IC: I added script unit_tesp.py. This scripts loads a sphere, measures its initial diameter, it increases the diameter of the sphere by factor 1.5, measures its new diameter, loads the fluid, starts integration for 5000 time steps - during the integration the sphere shrinks back almost to its original diameter, at the end the script measures the relaxed diameter, computes relative wrror of relaxed diameter to initial diameter, if error is under 0.1 percent it outputs is "Test passed", otherwise "Test failed." The output could be made more appropriate for automatic testing, but I do not know how.

Rudolf: The current documentation (doc/sphinx/oif.rst) does not reflect the pep8-naming now applied in the oif module. Also, some name changes (stretching -> resize) are not reflected. Can you please adapt the docs
Response IC: done

Rudolf: Also, the docs in doc/sphinx/inter.rst need updating (membrane_collision, ...). Currently, the TCL stat is described.
Response IC: done

RudolfWeeber committed:
Merge branch 'python' of git://github.com/espressomd/espresso into oif_merge

RudolfWeeber committed:
rm extra file

RudolfWeeber committed:
oif: Move test case to testsuite/ and convert to unit_test framework

RudolfWeeber committed:
oif: convert max_oif_objects from a macro to a global

toi avoid mpi communication for non existing soft objects

RudolfWeeber commented:
@fweik, from my side, it's ready.

I changed the maximum number of oif objects from a c macro to a global and gave it a Python interface via system.max_oif_objects.
The value is incremented when instancing an OifCell().
In this way, I could include the featuers into maxset without large performance impacts. So, the oif volume/area conservation is in the CI testing now.
@icimrak, I took LB out of your testcase, because the volume conservation is really independent from that. One only needs a drag force, which Langevin can deliver. The test takes 4 seconds, now.

Also, the current code performs an mpi allreduce and a sweep over all particles per soft object, if I understand the code correctly. 
That may result in a significant perofrmance drag if you have more than 2-3 soft objects. In the long run, you could consider collecting all the volumes in a single sweep in a std::vector<double> and mpi_allreducing that once.

fweik commented:
@KonradBreitsprecher @RudolfWeeber  Do you have a suggestion where the OIF docs should go in the new structure? It's not clear to me where to put this, see also the merge conflicts.

KonradBreitsprecher commented:
I would suggest that oif.rst is moved into a new section in 'Advanced Methods'.
The parts in inter.rst should go to their respective sections:

- Membrane-collision interaction 
-> Non-bonded interaction | Anisotropic non-bonded 

- OIF local forces, OIF global forces, Volume conservation, Out direction 
-> Bonded interaction | OIF Bonds (new section)

RudolfWeeber committed:
Merge branch 'python' of git://github.com/espressomd/espresso into oif_merge

RudolfWeeber commented:
@icimrak, could you please check that all your updated docs are in correctly? I had to move stuff around quite a bit due to the updated doc structure

The non-bonded (membrane collision) part is with non bonded interactions
The remaining oif interactions are with bonded interactions.
The description of the method is in advanced methods.

Commit:
Corrections in documentation fo OIF, sectioning corrected.

icimrak commented:
@RudolfWeeber 
I corrected sectioning levels in advanced_methods.rst in the part about OIF.

The rest looks good. 

One minor issue in file system_setup.rst, the error in compiling arose about reference ":ref:`Hat interaction`" on line 370. I didn't know how to correct it, it is outside OIF.

RudolfWeeber committed:
Doc: Re-added Hat ia which got lost in a merge

RudolfWeeber committed:
Merge branch 'python' of git://github.com/espressomd/espresso into oif_merge

RudolfWeeber commented:
@fweik, from my side its ready.

KaiSzuttor commented on doc/sphinx/advanced_methods.rst:
```diff
@@ -397,711 +397,1003 @@ ibm\_triel, ibm\_tribend and ibm\_volCons:
+  build directory before running cmake.
+  
+| Note: Do not change the myconfig\_default.hpp in the source directory,
+  since this file is created automatically by and is used only if no
```
by what?

icimrak replied:
Comment deleted, not OIF specific

KaiSzuttor commented on doc/sphinx/advanced_methods.rst:
```diff
@@ -397,711 +397,1003 @@ ibm\_triel, ibm\_tribend and ibm\_volCons:
-elastic modulus for local area forces
 
-elastic modulus for global area forces
+To check, what features are currently compiled into your , use the
```
into your what?

icimrak replied:
Comment deleted, not OIF specific

KaiSzuttor commented on doc/sphinx/advanced_methods.rst:
```diff
@@ -397,711 +397,1003 @@ ibm\_triel, ibm\_tribend and ibm\_volCons:
+command::
 
-elastic modulus for volume forces
+    code_info.features()
```
isnt it ``espressomd.features()``?

icimrak replied:
Comment deleted, not OIF specific

KaiSzuttor commented on doc/sphinx/advanced_methods.rst:
```diff
@@ -397,711 +397,1003 @@ ibm\_triel, ibm\_tribend and ibm\_volCons:
 
-   .. math:: ks_i = ksMin * (1 - weight_i) + ksMax*(weight_i)
+The script described in this section is available at
+http://cell-in-fluid.kst.fri.uniza.sk/en/content/oif-espresso. 
```
Shouldn’t the file reside in the samples folder?

icimrak replied:
File added to samples, text adapted....

KaiSzuttor commented on doc/sphinx/advanced_methods.rst:
```diff
@@ -397,711 +397,1003 @@ ibm\_triel, ibm\_tribend and ibm\_volCons:
+It can also be done by setting the velocity of the individual
+``lbnodes`` on one side of the channel and letting the flow develop, but
+this is not a very convenient setup because it has to be done at every
+integration step and the tcl-C communication slows down the computation.
```
tcl?

icimrak replied:
Comment deleted, because moving the fluid with LB nodes has almost never been used in OIF

KaiSzuttor reviewed (changes requested):
I think documentation should get another read by the authors

Commit:
Minor changes to documentation for OIF module

icimrak commented:
I went through the documentation and corrected some facts/mistypes here and there, Thanks Kai for pointing this out

Commit:
more user friendly VTK output for cylinder, rhomboid; added output for pore; corrected cell_type info output

Commit:
unified parameter names for oif_membrane_collision and oif_cells

icimrak committed:
Merge branch 'python' into python

Commit:
Merge remote-tracking branch 'ijancigova/python' into python

RudolfWeeber committed:
Merge branch 'python' of git://github.com/espressomd/espresso into oif_merge

RudolfWeeber committed:
Adapt test to current syntax

RudolfWeeber committed:
oif: Adapt samples

RudolfWeeber committed:
oif: more test fix

RudolfWeeber commented:
I merged the current Python brnach in preparation of the merge.
However, there seem to be interface changes in the object in fluid python module.
@icimrak, could you please adapt the scripts in the smples folder?
From our side, the box_l is now required as a keyword arg when instanceing the System class. That change is already applied to the sample and testcase scripts

Commit:
- sample scripts in samples/object-in-fluid adapted to the interface changes
- sample script two_cells.py deleted since it was just the same simulation as in motivation.py

icimrak commented:
sample scripts adapted
Ivan

RudolfWeeber commented:
Looks like you didn't add motivation.py to the commit. Could you please re-check?
Thx

Commit:
added missing file motivation.py

icimrak commented:
Sorry, missing file added....

RudolfWeeber committed:
Merge branch 'python' of git://github.com/espressomd/espresso into oif_merge

RudolfWeeber committed:
Testsuite: Increase EK integration length for one_species_x

RudolfWeeber committed:
Merge branch 'python' of git://github.com/espressomd/espresso into oif_merge

RudolfWeeber commented:
Clang static analysis is unhappy because of something in the hdf5 library:
/builds/espressomd/espresso/libs/h5xx/h5xx/attribute/attribute.hpp:48:5: error: operator=() should return 'attribute&' [misc-unconventional-assign-operator,-warnings-as-errors]
    attribute const& operator=(attribute other);
    ^
/builds/espressomd/espresso/libs/h5xx/h5xx/attribute/attribute.hpp:145:1: error: operator=() should return 'attribute&' [misc-unconventional-assign-operator,-warnings-as-errors]
inline attribute const& attribute::operator=(attribute other)
^
And several more of the same kind.

What do we do about this? Do we disable checking for the external library?

icimrak commented:
I have no idea, what could possibly be wrong. If I understand it correctly, hdf5 is a file format for storing large amount of data. In OIF, we have not written any output methods for storing in hdf5 format. So from our side, it is ok to disable external library...

KaiSzuttor commented:
i wonder why we dont have the errors for other PRs

RudolfWeeber committed:
Fix libs/h5xx commit id (was locked ot outdated commit)

RudolfWeeber commented:
Outdated git submodule commit leaked in for h5xx

fweik commented:
I'm strongly for merging this now, @KaiSzuttor can we do that?

fweik commented:
Thank you all very much for your contributions. I'm sorry that it took so long, but I think we came to a good result.
